code/modules/vehicles/tank.dm

```diff
@@ -0,0 +1,381 @@
+		return TRUE
+
+/obj/vehicle/tank/proc/enter(var/mob/user, var/position) //By this point, we've checked that the seats are actually empty, so we won't need to do that again HOPEFULLY
+	if(!user || !position)
+		return //Something fucked up. Whoops!
+	user.forceMove(src)
+	operators += user
+	switch(position)
+		if("pilot")
+			pilot = user
+		if("gunner")
+			gunner = user
+
+
+/obj/vehicle/tank/proc/remove_all_players()
+	for(var/M in operators)
+		exit_tank(M)
+
+/obj/vehicle/tank/proc/exit_tank(var/mob/user) //By this point, we've checked that the seats are actually empty, so we won't need to do that again HOPEFULLY
+	if(!user)
+		return //Something fucked up. Whoops!
+	if(user == pilot)
+		user.forceMove(get_turf(src))
+		pilot = null
+		operators -= user
+	if(user == gunner)
+		user.forceMove(get_turf(src))
+		gunner = null
+		operators -= user
+
+/obj/vehicle/tank/relaymove(mob/user, direction)
+	if(user.incapacitated() || user != pilot)
+		to_chat(user, "You can't reach the gas pedal from down here, maybe try manning the driver's seat?")
+		return
+	if(world.time > last_move_time + move_delay)
+		. = step(src, direction)
+	update_icon()
+
+/obj/vehicle/tank/Bump(atom/A, yes)
+	. = ..()
+	var/facing = get_dir(src, A)
+	var/turf/temp = get_turf(loc)
+	var/turf/T = get_turf(loc)
+	A.tank_collision(src, facing, T, temp)
+	if(isliving(A) && pilot)
+		log_attack("[pilot] drove over [A] with [src]")
+
+/client/MouseDown(object, location, control, params)
+	. = ..()
+	if(istype(mob.loc, /obj/vehicle/tank)) //mob is a var that exists on all clients anyway
+		var/obj/vehicle/tank/our_tank = mob.loc //I could refactor this to be a var on the mob, if you'd prefer
+		our_tank.onMouseDown(object,mob,params) //Alright, so as soon as they start clicking, this means they want to start firing. If we're in a tank that means that they want to shoot the tank gun
+		//MAINTAINERS: You can easily refactor this to work with guns, just follow the above format.
+
+/client/MouseUp(object, location, control, params) ///Oh boy click code, my favourite! This adds support for click n' hold gun firing action
+	. = ..()
+	if(istype(mob.loc, /obj/vehicle/tank))
+		var/obj/vehicle/tank/our_tank = mob.loc //I could refactor this to be a var on the mob, if you'd prefer
+		our_tank.onMouseUp(object,mob)
+
+
+//Combat, tank guns, all that fun stuff
+/obj/vehicle/tank/proc/onMouseDown(var/atom/A, mob/user, params)
+	if(user != gunner) //Only the gunner can fire!
+		return
+	var/list/modifiers = params2list(params) //If they're CTRL clicking, for example, let's not have them accidentally shoot.
+	if(modifiers["shift"])
+		return
+	if(modifiers["ctrl"])
+		handle_fire_main(A) //CTRL click to fire your big tank gun you can change any of these parameters here to hotkey for other shit :)
+		return
+	if(modifiers["middle"])
+		return
+	if(modifiers["alt"])
+		return
+	handle_fire(A)
+
+/obj/vehicle/tank/proc/handle_fire(var/atom/A)
+	if(!minigun && gunner)
```

Comment:
If we use a signal for this, then checking for `minigun` and `gunner` is not needed.